adapters/adocean/adocean.go

```diff
@@ -0,0 +1,324 @@
+			endpointURL.RawQuery = queryParams.Encode()
+			newUri := endpointURL.String()
+			if len(newUri) < maxUriLength {
+				request.Uri = newUri
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (a *AdOceanAdapter) makeURL(params *openrtb_ext.ExtImpAdOcean, auctionID string, request *openrtb.BidRequest, consentString string) (string, error) {
+	if error := validateParams(params); error != nil {
+		return "", error
+	}
+
+	endpointParams := macros.EndpointTemplateParams{Host: params.EmitterDomain}
+	host, err := macros.ResolveMacros(a.endpointTemplate, endpointParams)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Unable to parse endpoint url template: " + err.Error(),
+		}
+	}
+
+	endpointURL, err := url.Parse(host)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Malformed URL: " + err.Error(),
+		}
+	}
+
+	randomizedPart := 10000000 + rand.Intn(99999999-10000000)
+	if request.Test == 1 {
+		randomizedPart = 10000000
+	}
+	endpointURL.Path = "/_" + strconv.Itoa(randomizedPart) + "/ad.json"
+
+	queryParams := url.Values{}
+	queryParams.Add("pbsrv_v", adapterVersion)
+	queryParams.Add("id", params.MasterID)
+	queryParams.Add("nc", "1")
+	queryParams.Add("nosecure", "1")
+	queryParams.Add("aid", params.SlaveID+":"+auctionID)
+	if consentString != "" {
+		queryParams.Add("gdpr_consent", consentString)
+		queryParams.Add("gdpr", "1")
+	}
+	if request.User != nil && request.User.BuyerUID != "" {
+		queryParams.Add("hcuserid", request.User.BuyerUID)
+	}
+	endpointURL.RawQuery = queryParams.Encode()
+
+	return endpointURL.String(), nil
+}
+
+func validateParams(params *openrtb_ext.ExtImpAdOcean) error {
+	if params.EmitterDomain == "" {
+		return &errortypes.BadInput{
+			Message: "Emitter domain undefined",
+		}
+	}
+
+	if params.MasterID == "" {
+		return &errortypes.BadInput{
+			Message: "MasterId undefined",
+		}
+	}
+
+	if params.SlaveID == "" {
+		return &errortypes.BadInput{
+			Message: "SlaveId undefined",
+		}
+	}
+
+	return nil
+}
+
+func (a *AdOceanAdapter) MakeBids(
+	internalRequest *openrtb.BidRequest,
+	externalRequest *adapters.RequestData,
+	response *adapters.ResponseData,
+) (*adapters.BidderResponse, []error) {
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Network error?", response.StatusCode)}
+	}
+
+	requestURL, _ := url.Parse(externalRequest.Uri)
+	queryParams := requestURL.Query()
+	auctionIDs := queryParams["aid"]
+
+	bidResponses := make([]ResponseAdUnit, 0)
+	if err := json.Unmarshal(response.Body, &bidResponses); err != nil {
+		return nil, []error{err}
+	}
+
+	var parsedResponses = adapters.NewBidderResponseWithBidsCapacity(len(auctionIDs))
+	var errors []error
+
+	for _, auctionFullID := range auctionIDs {
+		auctionIDsSlice := strings.SplitN(auctionFullID, ":", 2)
+		auctionID := auctionIDsSlice[1]
+		slaveID := auctionIDsSlice[0]
+
+		for _, bid := range bidResponses {
+			if bid.ID == slaveID {
```

Comment:
Small optimization. If it makes sense in the context of this adapter, we could go from the nested for loops of O(nm) time to O(n+m) consecutive loops at the cost of a little memory (currently, `prebid-server` can spare a bit more memory in return of less processing time). Again, this is only a suggestion and I'm ok if you think it doesn't make sense in this context.
```
252   func (a *AdOceanAdapter) MakeBids(
253       internalRequest *openrtb.BidRequest,
254       externalRequest *adapters.RequestData,
255       response *adapters.ResponseData,
256   ) (*adapters.BidderResponse, []error) {
257   *-- 15 lines: if response.StatusCode != http.StatusOK {---------------------
272  
    +     idMap := new(map[string]string,len(auctionIDs))
273       for _, auctionFullID := range auctionIDs {
274           auctionIDsSlice := strings.SplitN(auctionFullID, ":", 2)
275 -         auctionID := auctionIDsSlice[1]
276 -         slaveID := auctionIDsSlice[0]
    +         idMap[auctionIDsSlice[0]] = auctionIDsSlice[1]
    +     }
277  
278           for _, bid := range bidResponses {
279 -             if bid.ID == slaveID {
    +             if auctionID, found := idMap[bid.ID]; found {
280                   if bid.Error == "true" {
281                       break
282                   }
283  
284   *--  21 lines: price, _ := strconv.ParseFloat(bid.Price, 64)----------------
305                   parsedResponses.Currency = bid.Currency
306  
307                   break
308               }
309           }
310       }
311  
312       return parsedResponses, errors
313   }
adapters/adocean/adocean.go
```

Reply:
I don't expect number of slaves to be high enough to make a visible difference, but your version is also easier to read so I've changed it to your version, thanks :).